lib/usd/translators/shading/mtlxBaseReader.cpp

```diff
@@ -152,4 +140,79 @@ bool MtlxUsd_BaseReader::GetColorAndAlphaFromInput(
     return false;
 }
 
+bool MtlxUsd_BaseReader::TraverseUnconnectableInput(const TfToken& usdAttrName)
+{
+    TfToken               usdPortName;
+    UsdShadeAttributeType attrType;
+    std::tie(usdPortName, attrType) = UsdShadeUtils::GetBaseNameAndType(usdAttrName);
+
+    if (attrType == UsdShadeAttributeType::Output) {
+        return false;
+    }
+
+    // Check for the presence of a CTOR node indicating connection to a subcomponent of a compound
+    // input:
+    const auto&    prim = _GetArgs().GetUsdPrim();
+    UsdShadeShader shaderSchema = UsdShadeShader(prim);
+
+    UsdShadeInput usdInput = shaderSchema.GetInput(usdPortName);
+    if (usdInput) {
+        UsdShadeConnectableAPI source;
+        TfToken                sourceInputName;
+        UsdShadeAttributeType  sourceType;
+        if (UsdShadeConnectableAPI::GetConnectedSource(
+                usdInput, &source, &sourceInputName, &sourceType)) {
+
+            // Surface nodes will connect to a NodeGraph. We must dig deeper in this case:
+            UsdShadeNodeGraph nodeGraph(source.GetPrim());
+            if (nodeGraph) {
+                UsdShadeOutput graphOutput = source.GetOutput(sourceInputName);
+                if (!graphOutput) {
+                    // Not a NodeGraph we recognize.
+                    false;
+                }
+                UsdShadeConnectableAPI::GetConnectedSource(
+                    graphOutput, &source, &sourceInputName, &sourceType);
+            }
+
+            return _IsConstructorNode(source.GetPrim());
+        }
+    }
+
+    return false;
+}
+
+void MtlxUsd_BaseReader::RegisterConstructorNodes(
+    UsdMayaPrimReaderContext& context,
+    MObject                   mayaObject)
+{
+    const auto&    prim = _GetArgs().GetUsdPrim();
+    UsdShadeShader shaderSchema = UsdShadeShader(prim);
+
+    for (auto&& usdInput : shaderSchema.GetInputs()) {
+        UsdShadeConnectableAPI source;
+        TfToken                sourceInputName;
+        UsdShadeAttributeType  sourceType;
+        if (UsdShadeConnectableAPI::GetConnectedSource(
+                usdInput, &source, &sourceInputName, &sourceType)) {
+
+            // Surface nodes will connect to a NodeGraph. We must dig deeper in this case:
+            UsdShadeNodeGraph nodeGraph(source.GetPrim());
+            if (nodeGraph) {
+                UsdShadeOutput graphOutput = source.GetOutput(sourceInputName);
+                if (!graphOutput) {
+                    // Not a NodeGraph we recognize.
+                    continue;
+                }
+                UsdShadeConnectableAPI::GetConnectedSource(
+                    graphOutput, &source, &sourceInputName, &sourceType);
+            }
+
+            if (_IsConstructorNode(source.GetPrim())) {
+                context.RegisterNewMayaNode(source.GetPath().GetString(), mayaObject);
```

Comment:
Registers the constructor nodes. Traversing UsdShade connections can only be done one-way, so a constructor node can not directly discover what it is writing to. This allows a constructor node to know that.